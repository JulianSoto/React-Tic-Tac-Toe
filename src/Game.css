.game-container {
	display: flex;
	grid-template-columns: auto auto;
}