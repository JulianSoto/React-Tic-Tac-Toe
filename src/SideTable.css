.side-table {
	position: relative;
	width: 150px;
	height: 150px;
	margin-left: 50px;
}

.status {
	position: absolute;
	bottom: 0;
	left: 0;
	width: 100%;
}

.home {
	width: 100%;
}

.game-mode-option {
	width: 100%;
}

.play-again {
	width: 100%;
}

.scoreboard {
	display: block;
	margin-top: 10px;
	font-size: 1.5em;
	color: #444;
}