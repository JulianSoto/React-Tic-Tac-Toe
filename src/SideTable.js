//side table
import React, {Component} from 'react';
import {HUMAN_VS_HUMAN, HUMAN_VS_COMPUTER} from './globalConstants.js';
import Status from './StatusMessage.js';

export default class SideTable extends Component {
	constructor(props){
		super(props);
	}

	render(){
		if (this.props.gameMode === null){
			return (
				<div>
					<button
						onClick={() => this.props.gameModeSelectionHandler(HUMAN_VS_HUMAN)}
					>
						Human vs human
					</button>
					<button
						onClick={() => this.props.gameModeSelectionHandler(HUMAN_VS_COMPUTER)}
					>
						Human vs computer
					</button>
					<span>Choose a game mode</span>
				</div>
			)
		}

		if (!this.props.players){
			return (
				<div>
					<button onClick={() => this.props.resetGameState()}>Home</button>
					<button onClick={() => this.props.selectX()}>X</button>
					<button onClick={() => this.props.selectO()}>O</button>
					<span>Select X or O</span>
				</div>
			)
		}

		return (
			<div>
				<button onClick={() => this.props.resetGameState()}>Home</button>
				<button>Reset</button>
				<Status
					winner={this.props.winner}
					players={this.props.players}
					isDraw={this.props.isDraw}
					gameMode={this.props.gameMode}
					turn={this.props.turn}
				/>
				<span>
					{`${this.props.scoreboard[0]} - ${this.props.scoreboard[1]}`}
				</span>
			</div>
		)
	}
}