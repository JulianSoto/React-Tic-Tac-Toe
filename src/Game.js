import React, { Component } from 'react';
import './Game.css';
import SideTable from './SideTable.js';

import Board from './Board';

const initialState = {
	turn: 0, 		//turn's value represents player's index
	players: null, 	//computer's index is always 1
	winner: null,
	isDraw: false,
	gameMode: null,
	squares: Array(9).fill(null),
	scoreboard: [0, 0]
};

class Game extends Component {
	constructor(props){
		super(props);
		this.state = initialState;
		this.setGameMode = this.setGameMode.bind(this);
		this.resetGameState = this.resetGameState.bind(this);
	}

	setGameMode(mode){
		this.setState({
			gameMode: mode
		});
	}

	resetGameState(){
		this.setState(initialState)
	}

	render() {
		return (
			<div>
				<SideTable
					className="side-table"
					winner={this.state.winner}
					isDraw={this.state.isDraw}
					turn={this.state.turn}
					players={this.state.players}
					scoreboard={this.state.scoreboard}
					gameMode={this.state.gameMode}
					gameModeSelectionHandler={this.setGameMode}
					resetGameState={this.resetGameState}
				/>
				{
					this.state.players ?
						<Board 
							className="board"
							squares={this.state.squares}
						/>
					: null
				}
			</div>
		);
	}
}

export default Game;
