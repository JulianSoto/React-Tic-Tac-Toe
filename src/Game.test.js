import React from 'react';
import { shallow } from 'enzyme';
import SideTable from './SideTable.js';

import Game from './Game';

describe('Game component', () => {
	it('expect to match snapshot', () => {
		const wrapper = shallow(<Game />);
		expect(wrapper).toMatchSnapshot();
	});

	it('Should contain a side table', () => {
		const wrapper = shallow(<Game />);
		expect(wrapper.find('.side-table').exists()).toBe(true);
	});

	it(`Shouldn't contain a board`, () => {
		const wrapper = shallow(<Game />);
		expect(wrapper.find('.board').exists()).toBe(false);
	});

	it(`Should contain a board when players is set in state`, () => {
		const wrapper = shallow(<Game />);
		wrapper.setState({players: ['X', 'O']});
		expect(wrapper.find('.board').exists()).toBe(true);
	});
});