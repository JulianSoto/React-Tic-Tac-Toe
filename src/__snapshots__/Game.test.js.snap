// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game component expect to match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Game />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <SideTable
          className="side-table"
          gameMode={null}
          gameModeSelectionHandler={[Function]}
          isDraw={false}
          players={null}
          resetGameState={[Function]}
          scoreboard={
            Array [
              0,
              0,
            ]
          }
          turn={0}
          winner={null}
        />,
        null,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "side-table",
          "gameMode": null,
          "gameModeSelectionHandler": [Function],
          "isDraw": false,
          "players": null,
          "resetGameState": [Function],
          "scoreboard": Array [
            0,
            0,
          ],
          "turn": 0,
          "winner": null,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <SideTable
            className="side-table"
            gameMode={null}
            gameModeSelectionHandler={[Function]}
            isDraw={false}
            players={null}
            resetGameState={[Function]}
            scoreboard={
              Array [
                0,
                0,
              ]
            }
            turn={0}
            winner={null}
          />,
          null,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "side-table",
            "gameMode": null,
            "gameModeSelectionHandler": [Function],
            "isDraw": false,
            "players": null,
            "resetGameState": [Function],
            "scoreboard": Array [
              0,
              0,
            ],
            "turn": 0,
            "winner": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
